experimental/pgbouncer_with_vault/install.sh

```diff
@@ -8,21 +8,29 @@ CONSUL_NAME="consul"
 VAULT_NAME="vault"
 PGXL_SERVICE_NAME="db-vlt-pgb-postgres-xl-svc"
 
+#=================================================================================================
+# REUSABLE FUNCTIONS
+#-------------------------------------------------------------------------------------------------
+source ./functions.sh
+#=================================================================================================
+
 #=================================================================================================
 # SETUP PGXL
 #-------------------------------------------------------------------------------------------------
-BASE64_PASSWORD=$(printf "${PASSWORD}" | base64)
+BASE64_PASSWORD=$(printf "%s" "${PASSWORD}" | base64)
 
-echo "apiVersion: v1
+SECRET_TEMPLATE=$'apiVersion: v1
```

Comment:
This can be in a file as well...

If you move the yaml to a file,  then,

```shell
cat [template filename] | xargs replace_with_env | kubectl apply -f -
```

This way there is a complete seperation between scripts, yaml and sql. 

We can skip this if you think this is too much for experimental, but I think it would give clarity.

Reply:
The same can be done for all the other scripts in the system.

Reply:
ok nice I'll do that.

Reply:
When I try do this I get no such file or directory error like this https://stackoverflow.com/questions/11003418/calling-shell-functions-with-xargs still there with the workarounds. Have you ever had this issue before?

Reply:
I have tested this localy.. seems to work. 
Can you send out the whole error?

Reply:
Hi, 

Misconception on my end, you cannot pipe to a function unless you do lie [so.](https://unix.stackexchange.com/questions/114121/how-to-compose-bash-functions-using-pipes)

For our case, the following works:

pod.yaml:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: {{POD_NAME}}
  labels:
    name: {{POD_NAME}}
spec:
  containers:
  - name: {{POD_NAME}}
    image: {{POD_IMAGE}}
    resources:
      limits:
        memory: "128Mi"
        cpu: "500m"
    ports:
      - containerPort: {{POD_PORT}}
```

functions.sh:
```shell
#!/usr/bin/env bash

function replace_with_env(){
    # replace any {{ENV_NAME}} with its respective env value.
    value="$1"

    # the regular expression matches {{ [SOME NAME] }}
    re='(.*)\{\{\s*(\w+)\s*\}\}(.*)'

    # search for all replacements.
    while [[ "$value" =~ $re ]]; do
        env_name=${BASH_REMATCH[2]}
        env_value=${!env_name}
        value="${BASH_REMATCH[1]}$env_value${BASH_REMATCH[3]}"
    done

    echo "$value"
}```

main:
```shell
#!/usr/bin/env bash

# load methods
source "./functions.sh"

# set template values.
export POD_NAME="lama"
export POD_IMAGE="ubuntu"
export POD_PORT="3333"

# create the config and apply.
YAML_CONFIG=$(replace_with_env "$(cat ./pod.yaml)")
echo "$YAML_CONFIG"
echo "$YAML_CONFIG" | kubectl apply -f -
```

Which will create the pod and print:
```shell
apiVersion: v1
kind: Pod
metadata:
  name: lama
  labels:
    name: lama
spec:
  containers:
  - name: lama
    image: ubuntu
    resources:
      limits:
        memory: "128Mi"
        cpu: "500m"
    ports:
      - containerPort: 3333
pod/lama created
```